File: android/src/main/java/com/margelo/nitro/nitroimagepipeline/HybridNitroImagePipeline.kt
package com.margelo.nitro.nitroimagepipeline

import android.content.Context
import android.graphics.Bitmap
import androidx.core.graphics.drawable.toBitmap
import androidx.core.net.toUri
import coil3.BitmapImage
import coil3.ColorImage
import coil3.DrawableImage
import coil3.ImageLoader
import coil3.decode.DecodeResult
import coil3.decode.Decoder
import coil3.disk.DiskCache
import coil3.memory.MemoryCache
import coil3.network.okhttp.OkHttpNetworkFetcherFactory
import coil3.request.CachePolicy
import coil3.request.ErrorResult
import coil3.request.ImageRequest
import coil3.request.SuccessResult
import coil3.request.allowHardware
import coil3.request.transformations
import coil3.size.Scale
import coil3.size.Size
import coil3.transform.RoundedCornersTransformation
import com.google.net.cronet.okhttptransport.CronetInterceptor
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.margelo.nitro.image.HybridImage
import com.margelo.nitro.image.HybridImageSpec
import com.margelo.nitro.nitroimagepipeline.transform.BlurTransformation
import com.margelo.nitro.nitroimagepipeline.transform.HardwareBitmapTransformation
import com.margelo.nitro.nitroimagepipeline.transform.ResizeTransformation
import kotlin.math.roundToInt
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okio.Path.Companion.toOkioPath
import org.chromium.net.CronetEngine

class HybridNitroImagePipeline : HybridNitroImagePipelineSpec() {
  private val context
    get() = NitroModules.applicationContext as Context

  private val imageLoader: ImageLoader
    get() = getOrCreateImageLoader(context)

  init {
    // Build the shared loader now, on a background thread, instead of on
    // whichever thread the first request happens to run on. Creating it
    // builds a CronetEngine, an OkHttpClient and the disk cache — well over
    // a frame's worth of work — and the first `PipelineImageLoader` request
    // would otherwise do that on the main thread.
    (NitroModules.applicationContext as? Context)?.let { warmUpImageLoader(it) }
  }

  override fun loadImage(url: String, options: Options?): Promise<HybridImageSpec> = Promise.async {
    when (val result = imageLoader.execute(buildRequest(context, url, options))) {
      is ErrorResult -> throw result.throwable
      is SuccessResult -> HybridImage(bitmapOf(result))
    }
  }

  override fun createImageLoader(
      url: String,
      options: ViewOptions?,
  ): com.margelo.nitro.image.HybridImageLoaderSpec = PipelineImageLoader(context, url, options)

  // Preloading warms the disk cache only: the memory cache is skipped and the
  // decode step is replaced with a no-op (Coil's documented preload pattern),
  // so prefetching N URLs costs network + disk I/O instead of N full-resolution
  // bitmaps on the heap. The image is decoded — subsampled to the requested
  // size — only when a loadImage actually displays it.
  private fun preloadRequest(url: String): ImageRequest =
      ImageRequest.Builder(context)
          .data(requestData(context, url))
          .memoryCachePolicy(CachePolicy.DISABLED)
          .decoderFactory { _, _, _ -> Decoder { DecodeResult(ColorImage(), false) } }
          .build()

  override fun preLoadImage(url: String): Promise<Unit> = Promise.async {
    imageLoader.execute(preloadRequest(url))
  }

  override fun preLoadImages(urls: Array<String>): Promise<Unit> = Promise.async {
    coroutineScope {
      urls.map { url -> async { imageLoader.execute(preloadRequest(url)) } }.awaitAll()
    }
    Unit
  }

  override fun gaussianBlur(image: HybridImageSpec, radius: Double): Promise<HybridImageSpec> =
      Promise.async {
        val hybridImage = image as? HybridImage ?: throw Error("Image is not a HybridImage")
        // `radius` is a Gaussian sigma in source-image pixels — see BlurTransformation.
        val sigma = radius.toFloat()
        if (sigma <= 0f) {
          hybridImage
        } else {
          val blurred = BlurTransformation(sigma).transform(hybridImage.bitmap, Size.ORIGINAL)
          HybridImage(blurred)
        }
      }

  override fun setMemoryCacheLimit(bytes: Double) {
    require(bytes >= 0 && bytes.isFinite()) {
      "Memory cache limit must be a non-negative, finite number of bytes (got $bytes)"
    }
    imageLoader.memoryCache?.apply {
      maxSize = bytes.toLong()
      // Setting maxSize only affects future inserts; evict down to it now so
      // the call frees memory immediately.
      trimToSize(bytes.toLong())
    }
  }

  override fun clearCache(): Promise<Unit> = Promise.async {
    imageLoader.memoryCache?.clear()
    // DiskCache.clear() does file I/O; keep it off the JS thread but await
    // completion so callers can rely on the cache being empty, and so an
    // IOException rejects the promise instead of crashing the process.
    withContext(Dispatchers.IO) { imageLoader.diskCache?.clear() }
  }

  companion object {
    // Coil requires a single DiskCache instance per directory, so the loader
    // must be shared across all HybridNitroImagePipeline instances.
    @Volatile private var sharedImageLoader: ImageLoader? = null

    /**
     * The [ImageRequest] for [url] with [options] applied — shared by [loadImage] and
     * [PipelineImageLoader] so both hit the same caches with identical cache keys.
     *
     * With [hardwareResult], a transformed result is converted to a hardware bitmap before it is
     * cached (see [HardwareBitmapTransformation]). Only for callers that draw the result: its
     * pixels cannot be read back, and its cache key differs from the software one.
     */
    internal fun buildRequest(
        context: Context,
        url: String,
        options: Options?,
        hardwareResult: Boolean = false,
    ): ImageRequest {
      val blur = options?.blur?.toFloat() ?: 0f
      val resize =
          options?.resize?.let { r ->
            val width = r.width.roundToInt()
            val height = r.height.roundToInt()
            if (width > 0 && height > 0) width to height else null
          }
      val roundedCorners: RoundedCornersTransformation? =
          options
              ?.cornerRadius
              ?.match(
                  first = { radius ->
                    if (radius > 0.0) RoundedCornersTransformation(radius.toFloat()) else null
                  },
                  second = { radii ->
                    // RoundedCornersTransformation rejects negative radii; treat them as square.
                    val topLeft = (radii.topLeft?.toFloat() ?: 0f).coerceAtLeast(0f)
                    val topRight = (radii.topRight?.toFloat() ?: 0f).coerceAtLeast(0f)
                    val bottomLeft = (radii.bottomLeft?.toFloat() ?: 0f).coerceAtLeast(0f)
                    val bottomRight = (radii.bottomRight?.toFloat() ?: 0f).coerceAtLeast(0f)
                    if (topLeft > 0f || topRight > 0f || bottomLeft > 0f || bottomRight > 0f) {
                      RoundedCornersTransformation(topLeft, topRight, bottomLeft, bottomRight)
                    } else {
                      null
                    }
                  },
              )
      val transformations = buildList {
        // Resize first: blur sigma and corner radii are in pixels of the bitmap
        // they run on, so they must see the final size. Coil's
        // RoundedCornersTransformation already scale-fills and center-crops
        // to the request size (`resize`, when set) as part of its own draw,
        // so with rounded corners and no blur the explicit resize would only
        // add an intermediate bitmap — and the radii still apply to the final
        // size, because that is the size its output has.
        if (resize != null && (blur > 0f || roundedCorners == null)) {
          add(ResizeTransformation(resize.first, resize.second))
        }
        if (blur > 0f) add(BlurTransformation(blur))
        roundedCorners?.let { add(it) }
        // Without transformations Coil decodes straight to a hardware bitmap
        // already (allowHardware below); only a transformed result needs the
        // explicit upload.
        if (hardwareResult && isNotEmpty() && HardwareBitmapTransformation.isSupported) {
          add(HardwareBitmapTransformation.instance)
        }
      }
      return ImageRequest.Builder(context)
          .data(requestData(context, url))
          .apply {
            when (options?.cache) {
              CacheOption.MEMORY -> {
                memoryCachePolicy(CachePolicy.ENABLED)
                diskCachePolicy(CachePolicy.DISABLED)
              }
              CacheOption.DISK -> {
                memoryCachePolicy(CachePolicy.DISABLED)
                diskCachePolicy(CachePolicy.ENABLED)
              }
              CacheOption.NONE -> {
                memoryCachePolicy(CachePolicy.DISABLED)
                diskCachePolicy(CachePolicy.DISABLED)
              }
              null -> Unit // Coil defaults: both enabled
            }
            // Ask the decoder for the target size so a large source is
            // subsampled near it instead of decoded at full resolution;
            // ResizeTransformation then makes the size exact.
            resize?.let { (width, height) ->
              size(width, height)
              scale(Scale.FILL)
            }
          }
          .allowHardware(true)
          .transformations(transformations)
          .build()
    }

    /**
     * What Coil should load for [url]. Coil itself handles `http(s)://`, `file://`, `content://`
     * and plain absolute paths. The one form it doesn't is a bare resource name — a string with
     * no scheme, like `src_assets_logo` — which is what React Native's `require()` resolves to
     * in a release build (assets are packed into `res/drawable-*`). Resolve that to the resource
     * id here; the density-qualified variant is then picked by the resources system, like
     * `<Image>` does. An unknown name is passed through so Coil reports the failure as an
     * [ErrorResult] (the loaders must not throw while building a request).
     */
    internal fun requestData(context: Context, url: String): Any {
      if (url.startsWith("/") || url.toUri().scheme != null) return url
      // Tolerate `logo.png`: RN's generated names carry no extension, but nitro-image's
      // `{ resource: 'logo.png' }` convention does.
      val name = url.substringBefore('.')
      for (type in RESOURCE_TYPES) {
        val id = context.resources.getIdentifier(name, type, context.packageName)
        if (id != 0) return id
      }
      return url
    }

    /** Where React Native puts bundled image assets, then the other image resource types. */
    private val RESOURCE_TYPES = arrayOf("drawable", "mipmap", "raw")

    /** The decoded bitmap of a successful load. */
    internal fun bitmapOf(result: SuccessResult): Bitmap =
        when (val img = result.image) {
          is BitmapImage -> img.bitmap
          is DrawableImage -> img.drawable.toBitmap()
          else -> throw Error("Unsupported image type: ${img.javaClass.simpleName}")
        }

    internal fun getOrCreateImageLoader(context: Context): ImageLoader =
        sharedImageLoader
            ?: synchronized(this) {
              sharedImageLoader ?: createImageLoader(context).also { sharedImageLoader = it }
            }

    /**
     * The shared loader, creating it on [Dispatchers.IO] if it doesn't exist yet. Use this from
     * main-thread coroutines: once the loader exists this returns synchronously (so memory-cache
     * hits stay synchronous), and until then the expensive creation runs off the main thread.
     */
    internal suspend fun awaitImageLoader(context: Context): ImageLoader =
        sharedImageLoader ?: withContext(Dispatchers.IO) { getOrCreateImageLoader(context) }

    /** Starts creating the shared loader on a background thread, if it doesn't exist yet. */
    private fun warmUpImageLoader(context: Context) {
      if (sharedImageLoader != null) return
      CoroutineScope(Dispatchers.IO).launch {
        // Warm-up only: an uncaught exception here would take the process
        // down. If creation fails, `sharedImageLoader` stays null and the
        // next request retries it and surfaces the error itself.
        runCatching { getOrCreateImageLoader(context) }
      }
    }

    private fun createImageLoader(context: Context): ImageLoader {
      val okHttpClient =
          OkHttpClient.Builder()
              .apply {
                try {
                  val cronetEngine = CronetEngine.Builder(context).build()
                  addInterceptor(CronetInterceptor.newBuilder(cronetEngine).build())
                } catch (_: Exception) {
                  // Cronet unavailable (no GMS or unsupported device) — use plain OkHttp
                }
              }
              .build()

      return ImageLoader.Builder(context)
          .components { add(OkHttpNetworkFetcherFactory(callFactory = okHttpClient)) }
          .memoryCache { MemoryCache.Builder().maxSizePercent(context, 0.25).build() }
          .diskCache {
            DiskCache.Builder()
                .directory(context.cacheDir.resolve("nitro_image_cache").toOkioPath())
                .maxSizeBytes(256L * 1024 * 1024)
                .build()
          }
          .build()
    }
  }
}

File: ios/HybridNitroImagePipeline.swift
//
//  HybridNitroImagePipeline.swift
//  Pods
//
//  Created by Alejandro Paredes Alva on 3/31/2026.
//

import Foundation
import NitroModules
import NitroImage
import Nuke

import UIKit

class HybridImage: HybridImageSpec, NativeImage {
    let uiImage: UIImage

    // PNG encoding is expensive; encode once and reuse for both
    // toArrayBuffer() and toBase64(). Guarded by a lock (lazy var is not
    // thread-safe) and dropped on memory pressure so a JS-retained image
    // doesn't pin its encoding forever.
    private let pngLock = NSLock()
    private var cachedPngData: Data?
    // Registered only once there is encoded data to drop: most images are
    // only ever displayed, and an observer per image is wasted work in a
    // list of hundreds. Guarded by `pngLock` like the data it protects.
    private var memoryWarningObserver: (any NSObjectProtocol)?

    init(uiImage: UIImage) {
        self.uiImage = uiImage
        super.init()
    }

    deinit {
        if let memoryWarningObserver {
            NotificationCenter.default.removeObserver(memoryWarningObserver)
        }
    }

    private func pngData() -> Data? {
        pngLock.lock()
        defer { pngLock.unlock() }
        if let cachedPngData {
            return cachedPngData
        }
        let data = uiImage.pngData()
        cachedPngData = data
        if data != nil, memoryWarningObserver == nil {
            memoryWarningObserver = NotificationCenter.default.addObserver(
                forName: UIApplication.didReceiveMemoryWarningNotification,
                object: nil,
                queue: nil
            ) { [weak self] _ in
                guard let self else { return }
                self.pngLock.lock()
                self.cachedPngData = nil
                self.pngLock.unlock()
            }
        }
        return data
    }

    var width: Double {
        return Double(uiImage.size.width)
    }

    var height: Double {
        return Double(uiImage.size.height)
    }

    func toArrayBuffer() throws -> ArrayBuffer {
        guard let data = pngData() else {
            throw RuntimeError.error(withMessage: "Failed to encode image to PNG")
        }
        return try ArrayBuffer.copy(data: data)
    }

    func toBase64() throws -> String {
        guard let data = pngData() else {
            throw RuntimeError.error(withMessage: "Failed to encode image")
        }

        return data.base64EncodedString()
    }

    var description: String {
        return "HybridImage(\(width)x\(height))"
    }
}

class HybridNitroImagePipeline: HybridNitroImagePipelineSpec {
    // One shared pipeline (instead of mutating ImagePipeline.shared) so the
    // prefetcher and loadImage read/write the exact same caches, other Nuke
    // users in the host app are left untouched, and repeated instantiation
    // (e.g. Metro reloads) never puts two DataCache instances on the same
    // directory.
    static let sharedPipeline: ImagePipeline = {
        var configuration = ImagePipeline.Configuration.withDataCache
        // Store processed (blurred/rounded) variants on disk in addition to
        // the original download, so they survive memory eviction and
        // restarts without dropping the original for other variants.
        configuration.dataCachePolicy = .storeAll
        // A private, capped memory cache. The default (ImageCache.shared)
        // sizes itself at 15% of physical RAM — up to 768 MB of decoded
        // bitmaps on modern devices before anything is evicted, which shows
        // up as "high RAM usage" even though it is all evictable cache.
        // 128 MB still holds ~10 full-screen bitmaps or hundreds of list
        // thumbnails, and the LRU cache keeps trimming itself on memory
        // warnings and when the app enters the background.
        configuration.imageCache = ImageCache(
            costLimit: min(ImageCache.defaultCostLimit, 128 * 1024 * 1024)
        )
        // Nuke decodes on a serial queue by default, so a screenful of new
        // list cells is decoded one image at a time however many cores the
        // device has. Decode in parallel instead — capped so a burst of
        // full-resolution decodes (no `resize`) can't multiply peak memory
        // by the core count. Processing stays at Nuke's default of 2: the
        // blur allocates several full-size buffers per image.
        configuration.imageDecodingQueue.maxConcurrentOperationCount = min(
            4, max(1, ProcessInfo.processInfo.activeProcessorCount)
        )
        return ImagePipeline(configuration: configuration)
    }()

    // `.diskCache` stores the downloaded data without decoding it, so
    // prefetching N URLs costs network + disk I/O instead of N decoded
    // full-resolution bitmaps parked in the memory cache (the default
    // `.memoryCache` destination). The image is decoded — at the requested
    // target size — only when a loadImage actually displays it.
    private static let sharedPrefetcher = ImagePrefetcher(
        pipeline: sharedPipeline,
        destination: .diskCache
    )

    private var pipeline: ImagePipeline { Self.sharedPipeline }
    private var prefetcher: ImagePrefetcher { Self.sharedPrefetcher }

    private static func cacheOptions(for cache: CacheOption?) -> ImageRequest.Options {
        switch cache {
        case .memory: [.disableDiskCache]
        case .disk:   [.disableMemoryCache]
        case .none?:  [.disableDiskCache, .disableMemoryCache]
        default:      []
        }
    }

    /// The processor that bakes `cornerRadius` into the bitmap, or `nil` when
    /// no corner is actually rounded.
    private static func cornerRadiusProcessor(
        for cornerRadius: Variant_Double_CornerRadii?
    ) -> (any ImageProcessing)? {
        switch cornerRadius {
        case .first(let radius):
            guard radius > 0 else { return nil }
            // `unit: .pixels` is required: Nuke defaults to `.points`, which
            // multiplies the radius by the screen scale. The radius is
            // documented — and implemented on Android and in
            // RoundedCornersProcessor — as bitmap pixels.
            return ImageProcessors.RoundedCorners(radius: radius, unit: .pixels)
        case .second(let radii):
            let roundedCorners = RoundedCornersProcessor(radii: radii)
            return roundedCorners.hasRounding ? roundedCorners : nil
        case nil:
            return nil
        }
    }

    /// The target size in pixels, or `nil` when no (valid) resize was requested.
    private static func resizeSize(for options: Options?) -> CGSize? {
        guard let resize = options?.resize, resize.width > 0, resize.height > 0 else {
            return nil
        }
        return CGSize(width: resize.width, height: resize.height)
    }

    private static func processors(for options: Options?) -> [any ImageProcessing] {
        var processors: [any ImageProcessing] = []
        // Resize first: blur sigma and corner radii are defined in pixels
        // of the bitmap they run on, so they must see the final size.
        if let size = resizeSize(for: options) {
            processors.append(ImageProcessors.Resize(
                size: size,
                unit: .pixels,
                contentMode: .aspectFill,
                crop: true,
                upscale: true
            ))
        }
        if let blur = options?.blur, blur > 0 {
            processors.append(GaussianBlurProcessor(sigma: blur))
        }
        if let roundedCorners = cornerRadiusProcessor(for: options?.cornerRadius) {
            processors.append(roundedCorners)
        }
        return processors
    }

    /// The `ImageRequest` for `url` with `options` applied — shared by
    /// `loadImage` and `PipelineImageLoader` so both hit the same caches with
    /// identical cache keys.
    static func makeRequest(url: URL, options: Options?) -> ImageRequest {
        var imgRequest = ImageRequest(
            url: url,
            processors: processors(for: options),
            options: cacheOptions(for: options?.cache)
        )
        // With a target size known, decode near it (aspect-fill, so the
        // decoded image always covers the target) instead of at full
        // resolution — a 48 MP photo displayed as a 300 pt card would
        // otherwise decompress to ~190 MB before Resize shrinks it.
        // Matches Android, where the request's size() drives subsampling;
        // the exact size and crop still come from the Resize processor.
        if let size = resizeSize(for: options) {
            imgRequest.thumbnail = ImageRequest.ThumbnailOptions(
                size: size,
                unit: .pixels,
                contentMode: .aspectFill
            )
        }
        return imgRequest
    }

    func loadImage(url: String, options: Options?) throws -> Promise<any HybridImageSpec> {
        return Promise.async {
            guard let imageUrl = URL(string: url) else {
                throw RuntimeError.error(withMessage: "Invalid URL: \(url)")
            }

            let imgRequest = Self.makeRequest(url: imageUrl, options: options)
            let image = try await self.pipeline.image(for: imgRequest)
            return HybridImage(uiImage: image)
        }
    }

    func createImageLoader(url: String, options: ViewOptions?) throws -> any HybridImageLoaderSpec {
        // An unparseable URL fails at load time, not here — matching Android,
        // where Coil validates the URL only when the request runs.
        return PipelineImageLoader(url: url, options: options)
    }

    func preLoadImage(url: String) throws -> Promise<Void> {
        return Promise.async {
            guard let imageUrl = URL(string: url) else {
                throw RuntimeError.error(withMessage: "Invalid URL: \(url)")
            }

            self.prefetcher.startPrefetching(with: [imageUrl])
        }
    }

    func preLoadImages(urls: [String]) throws -> Promise<Void> {
        return Promise.async {
            let imageUrls = urls.compactMap { URL(string: $0) }
            self.prefetcher.startPrefetching(with: imageUrls)
        }
    }

    func setMemoryCacheLimit(bytes: Double) throws {
        guard bytes >= 0, bytes.isFinite else {
            throw RuntimeError.error(
                withMessage: "Memory cache limit must be a non-negative, finite number of bytes (got \(bytes))"
            )
        }
        guard let cache = pipeline.configuration.imageCache as? ImageCache else { return }
        // Double(Int.max) rounds up to 2^63, so Int(_:) of the clamped value
        // would trap; branch instead of min-ing.
        let limit = bytes >= Double(Int.max) ? Int.max : Int(bytes)
        cache.costLimit = limit
        // Setting the limit only affects future inserts; evict down to it now
        // so the call frees memory immediately.
        cache.trim(toCost: limit)
    }

    func clearCache() throws -> Promise<Void> {
        return Promise.async {
            self.pipeline.cache.removeAll()
        }
    }

    func gaussianBlur(image: any HybridImageSpec, radius: Double) throws -> Promise<any HybridImageSpec> {
        return Promise.async {
            guard let nativeImage = image as? NativeImage else {
                throw RuntimeError.error(withMessage: "Unsupported image type")
            }

            guard let blurred = GaussianBlur.apply(to: nativeImage.uiImage, sigma: radius) else {
                throw RuntimeError.error(withMessage: "Failed to apply gaussian blur")
            }

            return HybridImage(uiImage: blurred)
        }
    }

}
